### ISSUE TITLE
ML layers database

### ISSUE THREAD

**Issue by lberrada** [facebookresearch/TensorComprehensions#305]
Hi, I've noticed some examples that look incorrect to me in the ML layers database (both in the documentation and in the [tests](https://github.com/facebookresearch/TensorComprehensions/tree/master/test_python/layers)).
* Batch-normalization (issues: it is using running statistics to compute the output, and the running variance uses `rsqrt`). Suggestion of fix (result tested against pytorch):
```
def batchnorm(float(N,C,H,W) I, float(C) rMeanIn, float(C) rVarIn)
        -> (O, rMeanOut, rVarOut, centered, mean, variance)
        {{
           mean(c) +=! I(nn, c, hh, ww) / (N * H * W)
           centered(n, c, h, w) = I(n, c, h, w) - mean(c)
           variance(c) +=! (centered(n, c, h, w) * centered(n, c, h, w) + {eps}) / (N * H * W)
           O(n, c, h, w) = centered(n, c, h, w) * rsqrt(variance(c))
           rMeanOut(c) = (1 - {momentum}) * rMeanIn(c) + {momentum} * mean(c)
           rVarOut(c) = (1 - {momentum}) * rVarIn(c) + {momentum} * variance(c)
        }}
```
* Cosine similarity (issue: `rsqrt(sumI1(m))` instead of `sqrt(sumI1(m))`):
```
def cosine_similarity(float(M, N) I1, float(M, N) I2) -> (O, sumI1, sumI2) {{
    sumI1(m) +=! I1(m, n) * I1(m, n)
    sumI2(m) +=! I2(m, n) * I2(m, n)
    O(m) +=! (I1(m, n) * I2(m, n)) / fmax(sqrt(sumI1(m)) * sqrt(sumI2(m)), {eps})
}}
```
Let me know if you agree with these, and if you want me to make a PR.

**Comment by prigoyal** [comment]
Hi @lberrada , these sound great. Please feel free to send a PR :) your contribution is warmly welcomed.

**Comment by lberrada** [comment]
Ok thanks. I can include tests for most ML layers so that their output is compared against the pytorch implementation. Would that be something wanted?

**Comment by prigoyal** [comment]
@lberrada , yes, that will be awesome. Please feel free to send PR. > 95% TC users are using it via python so it will be a great service and contribution to the community. Thanks. :)

**Comment by ftynse** [comment]
@prigoyal as long as you are going to convert those into C++, #74 is still open even though you deprioritized it.

**Comment by ftynse** [comment]
Layer database no longer exists, we only use tests with actually executable code.
